library/std/src/alloc.rs

```diff
@@ -127,6 +127,122 @@ pub use alloc_crate::alloc::*;
 #[derive(Debug, Default, Copy, Clone)]
 pub struct System;
 
+use crate::sys::alloc::System as Imp;
+
+// When debug assertions are not enabled, `System` just forwards down to the particular platform
+// implementation.
+#[cfg(not(debug_assertions))]
+#[stable(feature = "alloc_system_type", since = "1.28.0")]
+unsafe impl GlobalAlloc for System {
```

Comment:
I do want to say that I think this is the right place to insert such checks, since it neatly sidesteps questions about what programs which override the global allocator get to rely on. Thanks!